gusto/equations.py

```diff
@@ -0,0 +1,168 @@
+        self.trial = TrialFunction(function_space)
+        self.bcs = None
+
+    def mass_term(self, q):
+        return inner(self.test, q)*dx
+
+    def add_term(self, term):
+        key = term.__class__.__name__
+        self.terms[key] = term
+
+
+class AdvectionEquation(Equation):
+    """
+    Class defining the advection equation.
+
+    Creates test and trial functions on the given function space,
+    defines the mass term and adds the advection term.
+    :arg function_space: :class:`.FunctionSpace` object. The function
+    space that the equation is defined on.
+
+    """
+    def __init__(self, function_space, state, **kwargs):
+
+        super().__init__(function_space)
+
+        self.add_term(AdvectionTerm(state, self.test, **kwargs))
+
+
+class ShallowWaterMomentumEquation(Equation):
+    """
+    Class defining the shallow water momentum equation.
+
+    Creates test and trial functions on the given function space,
+    defines the mass term and adds the shallow water momentum terms.
+    :arg function_space: :class:`.FunctionSpace` object. The function
+    space that the equation is defined on.
+    :arg state: :class:`.State` object.
+
+    """
+    def __init__(self, function_space, state):
+
+        super().__init__(function_space)
+
+        self.add_term(ShallowWaterPressureGradientTerm(state, self.test))
+        self.add_term(ShallowWaterCoriolisTerm(state, self.test))
+        self.add_term(VectorInvariantTerm(state, self.test))
+
+
+class ShallowWaterDepthEquation(AdvectionEquation):
+    """
+    Class defining the shallow water depth equation.
+
+    Creates test and trial functions on the given function space,
+    defines the mass term and adds continuity term.
+    :arg function_space: :class:`.FunctionSpace` object. The function
+    space that the equation is defined on.
+    :arg state: :class:`.State` object.
+
+    """
+    def __init__(self, function_space, state):
+
+        super().__init__(function_space, state, equation_form="continuity")
+
+
+class Equations(object, metaclass=ABCMeta):
+    """
+    Base equations class.
+
+    Builds compatible finite element function spaces as specified by
+    the child class.
+    Provides a list of prognostic fields, the mixed function space the
+    equations are defined on and when called with the name of a field,
+    will return the equation satisfied by that field.
+    :arg state: :class:`.State` object.
+    :arg family: string specifying the finite element family of the HDiv space
+    :arg degree: degree of the DG space
+
+    """
+
+    def __init__(self, state, family, degree):
+
+        self.state = state
+        self._build_function_spaces(state.spaces, state.mesh, family, degree)
+        state.fields = FieldCreator(self)
+
+        if hasattr(state, "diagnostics"):
+            state.diagnostics.register(*self.fieldlist)
+        else:
+            state.diagnostics = Diagnostics(*self.fieldlist)
```

Comment:
This ugliness could be avoided if `state` always made an (initially empty) `Diagnostics` object.